build_tools/azure/update_environments_and_lock_files.py

```diff
@@ -0,0 +1,412 @@
+    command = (
+        f"conda-lock lock --mamba --kind explicit --platform {platform} "
+        f"--file {environment_path} --filename-template {lock_file_path}"
+    )
+
+    logger.debug("conda-lock command: %s", command)
+    execute_command(shlex.split(command))
+
+
+def create_conda_lock_file(build_metadata, folder_path):
+    build_name = build_metadata["build_name"]
+    environment_path = folder_path / f"{build_name}_environment.yml"
+    platform = build_metadata["platform"]
+    lock_file_basename = build_name
+    if not lock_file_basename.endswith(platform):
+        lock_file_basename = f"{lock_file_basename}_{platform}"
+
+    lock_file_path = folder_path / f"{lock_file_basename}_conda.lock"
+    conda_lock(environment_path, lock_file_path, platform)
+
+
+def write_all_conda_lock_files(build_metadata_list, folder_path):
+    for build_metadata in build_metadata_list:
+        logger.info(build_metadata["build_name"])
+        create_conda_lock_file(build_metadata, folder_path)
+
+
+def get_pip_requirements_content(build_metadata):
+    template = environment.from_string(
+        """
+# DO NOT EDIT: this file is generated from the specification found in the
+# following script to centralize the configuration for all Azure CI builds:
+# build_tools/azure/update_environments_and_lock_files.py
+{% for pip_dep in build_metadata['pip_dependencies'] %}
+{{ pip_dep | get_package_with_constraint(build_metadata, uses_pip=True) }}
+{% endfor %}""".strip()
+    )
+    return template.render(build_metadata=build_metadata)
+
+
+def write_pip_requirements(build_metadata, folder_path):
+    build_name = build_metadata["build_name"]
+    content = get_pip_requirements_content(build_metadata)
+    output_path = folder_path / f"{build_name}_requirements.txt"
+    output_path.write_text(content)
+
+
+def write_all_pip_requirements(build_metadata_list, folder_path):
+    for build_metadata in build_metadata_list:
+        logger.info(build_metadata["build_name"])
+        write_pip_requirements(build_metadata, folder_path)
+
+
+def pip_compile(pip_compile_path, requirements_path, lock_file_path):
+    command = f"{pip_compile_path} {requirements_path} -o {lock_file_path}"
+
+    logger.debug("pip-compile command: %s", command)
+    execute_command(shlex.split(command))
+
+
+def write_pip_lock_file(build_metadata, folder_path):
+    build_name = build_metadata["build_name"]
+    python_version = build_metadata["python_version"]
+    environment_name = f"pip-tools-python{python_version}"
+    # To make sure that the Python used to create the pip lock file is the same
+    # as the one used during the CI build where the lock file is used, we first
+    # create a conda environment with the correct Python version and
+    # pip-compile and run pip-compile in this environment
+    command = (
+        "conda create -c conda-forge -n"
+        f" pip-tools-python{python_version} python={python_version} pip-tools -y"
```

Comment:
This command does not work for platforms where Python 3.8 does not exist, such as M1 macs. If there is no workaround, can we document this behavior in the file's docstring?

Reply:
OK I was not aware of this, I see some `osx-arm64` for Python 3.8 though on conda-forge e.g. [osx-arm64/python-3.8.13-hd3575e6_0_cpython.tar.bz2](https://anaconda.org/conda-forge/python/3.8.13/download/osx-arm64/python-3.8.13-hd3575e6_0_cpython.tar.bz2) though. Maybe `osx-arm64` is not the correct name for Apple M1 ...

I am open to suggestions on how to deal better with pure pip-based locks. This happens for Ubuntu for example where we want to use the system Python with apt-get installed packages (numpy, scipy, etc ...) and add a few packages on top of it (pytest, etc ...). 

We may be able to use Docker for this but that seems slightly overkill ... 

I tried to see whether there is a way to run pip-compile for another Python version, than the Python version in your current environment, but that does not seem to be the case according to https://github.com/jazzband/pip-tools#cross-environment-usage-of-requirementsinrequirementstxt-and-pip-compile.

Reply:
The current environment uses 3.8.3, which is not available on `osx-arm64`. The quick fix is to update the version to 3.8.13. The pins in the pip-compile lock file is not platform dependent, so it should be okay.